### ISSUE TITLE
Shadows break stereoscopic/3D rendering with post-processing

### ISSUE THREAD

**Issue by brianchirls** [mrdoob/three.js#7732]
There are certain cases where post-processing is necessary with stereoscopic rendering, e.g.:
- cardboard distortion effect when it's not supported natively by the browser in Web VR
- SSAO

Here's what currently happens in these cases, for each eye:
1. Set the render target to the buffer that will be post-processed
2. Set the scissor and [viewport to half the screen](https://github.com/mrdoob/three.js/blob/dev/examples/js/effects/VREffect.js#L148) (either left or right)
3. When `WebGLShadowMap.render` gets called, sets its own render target and viewport, but it stores the previously set viewport
4. When `WebGLShadowMap.render` is done, it [sets the viewport back to half screen](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLShadowMap.js#L298).
5. When `WebGLRenderer.render` continues, it [sets the render target again](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L1144), but `setRenderTarget` [sets the viewport to the full size of the render target](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L3405).
6. `WebGLRenderer` renders the scene at full size, across both eyes, breaking the output.

Somewhere at or after step 5, we need to be able to set the viewport back to the half screen values set in step 2.

While we're at it, as a bonus, it'd be nice to have a way to only render the shadow map one time. Repeating for the second eye is redundant.

(cc @borismus)

**Comment by borismus** [comment]
Thanks for logging this. Related to some issues folks have been filing against webvr-boilerplate: 

- https://github.com/borismus/webvr-boilerplate/issues/40
- https://github.com/borismus/webvr-boilerplate/issues/36

**Comment by brianchirls** [comment]
You're welcome, and thanks for sharing those other issues. I had a hard time believing I was the only one in the world who tried using shadows in Web VR. ;-)

**Comment by mrdoob** [comment]
Have you guys tried with the `dev` version? I fixed an issue with shadowmap and viewports the other day.

**Comment by brianchirls** [comment]
@mrdoob: Don't know, 'cause I can't get the `dev` version to run with webpack. `self` is undefined.

**Comment by mrdoob** [comment]
Oh well...

**Comment by brianchirls** [comment]
Is that something that's going to be fixed, or should I file a separate
issue for it?

**Comment by mrdoob** [comment]
I don't have experience using webpack. Hopefully someone that uses it figures out what the problem is and fixes it.

**Comment by brianchirls** [comment]
Okay, I'll file an issue and see what I can do to narrow down the problem.
Any idea if there's an issue or pull request that addresses why 'self' was
used in the first place?

Also, can you point to a commit with the changes that affect shadows and
viewport? I could test it separately and see if it solves the problem.

Thx

**Comment by mrdoob** [comment]
> Also, can you point to a commit with the changes that affect shadows and viewport?

Already did that some posts ago...

**Comment by brianchirls** [comment]
So you did. Sorry, I wrote that email from my phone and didn't see it. I'll take a look.

**Comment by brianchirls** [comment]
@mrdoob I'm afraid that commit did not fix the bug. The problem seems to be in step 5 (see above), which happens in `WebGLRenderer.render`.

**Comment by mrdoob** [comment]
I see... Do you mind providing a jsfiddle/example I can use to fix the issue?

**Comment by brianchirls** [comment]
Here you go: https://jsbin.com/kesodi/edit?js,output

I'm actually kinda curious to see what this'll look like in cardboard when it's working.

**Comment by mrdoob** [comment]
Shadows seem to work fine for that example on my Nexus 5. What's the issue?

![screenshot_20151211-010417](https://cloud.githubusercontent.com/assets/97088/11737104/cbc07b46-9fa3-11e5-943b-ee89af4a106b.png)

**Comment by brianchirls** [comment]
That's supposed to be stereoscopic. i.e. there should be one monolith in each eye. Try turning shadows off and then see what the output looks like.

**Comment by mrdoob** [comment]
Oh! LOL. Of course... 😅
OK, I'll continue investigating 😊

**Comment by brianchirls** [comment]
;-)

**Comment by mrdoob** [comment]
Ok, instead of going top to bottom I'm going bottom to top and see when it breaks.

http://jsfiddle.net/ebayetvL/

This works so far... I guess now we need to apply the Cardboard effect?

**Comment by brianchirls** [comment]
Yeah, the problem only shows up if you have a post-processing pass. The Cardboard distortion effect as included in my example will do it.

The easiest way to debug that is to use Chrome Dev Tools and enable "Device Mode". The Cardboard distortion effect has a build-in whitelist of phones that it works on, because it needs to know the physical screen dimensions. So you need to emulate the UA string of one of those devices to make it work. Nexus 5 will do, or you can see the full whitelist [here](https://github.com/borismus/webvr-boilerplate/blob/master/src/device-info.js).

**Comment by mrdoob** [comment]
Can we add the distortion effect to my jsfiddle instead? I would prefer not to have to dig inside `webvr-boilerplate`.

**Comment by brianchirls** [comment]
I think any post-processing effect should do it. I tried modifying your jsfiddle, but I'm hitting my head against the wall a bit here to make it work since I'm not super familiar with post-processing in three.js. (I usually [roll my own](https://github.com/brianchirls/Seriously.js/).)

The key point here is that the output of a stereo effect needs to be rendered to a render target, rather than directly to the screen. And then that render target needs to be passed through another effect and then rendered to the screen.

Another thing that might be causing some trouble/confusion here is that neither `VREffect` nor `StereoEffect` allows passing of a render target to the `render` method. (I was using a custom version of `VREffect` on some of my own projects that did that, and it's a pretty simple patch.) So webvr-boilerplate is forced to call `setRenderTarget` before calling `render`, which doesn't always work since `render` sets it again.

**Comment by mrdoob** [comment]
Success!

![screen shot 2015-12-19 at 13 37 07](https://cloud.githubusercontent.com/assets/97088/11915569/e01867ac-a660-11e5-99ba-17a81ca65899.png)

Unfortunately, `webvr_boilerplate` will need a few changes though.

https://github.com/mrdoob/three.js/blob/dev/examples/js/effects/CardboardEffect.js#L14
https://github.com/mrdoob/three.js/blob/dev/examples/js/effects/CardboardEffect.js#L94-L103

/ping @borismus

**Comment by brianchirls** [comment]
Wow, looks like you put a lot of work into this. Thanks. I will take a closer look at the code when I have more time. I would like to try this out with another post-processing effect, like maybe SSAO, just to be sure.

It might make things a little easier and more stable if the `render` method for `CardboardEffect`, `VREffect` and all the other stereo effects accepted a third argument of a `renderTarget`, just like the method on `WebGLRenderer`. I can do this myself and make a PR if you like.

**Comment by mrdoob** [comment]
> It might make things a little easier and more stable if the `render` method for `CardboardEffect`, `VREffect` and all the other stereo effects accepted a third argument of a `renderTarget`, just like the method on `WebGLRenderer`.

I actually wanted to remove `renderTarget` and `forceClear` from `.render()` 😁

**Comment by brianchirls** [comment]
Interesting. well, if it's robust and consistently works by setting `renderTarget` outside the `render` method, even with shadows, than that's great too.

**Comment by WestLangley** [comment]
>I actually wanted to remove renderTarget and forceClear from WebGLRenderer.render()

@mrdoob What is the new pattern you are thinking of?

**Comment by mrdoob** [comment]
I was thinking something like this:

```js
var renderTarget = new THREE.WebGLRenderTarget( renderer ):
renderTarget.setSize( 256, 256 );
renderTarget.render( scene, camera );
```

**Comment by mrdoob** [comment]
Works with PointLights too! ✌️

![screen shot 2015-12-19 at 19 35 49](https://cloud.githubusercontent.com/assets/97088/11916389/ef06893c-a689-11e5-9923-07e710c08b0c.png)

**Comment by mrdoob** [comment]
@borismus `THREE.Mirror` also works now.

![screen shot 2015-12-20 at 00 39 43](https://cloud.githubusercontent.com/assets/97088/11917216/a3b122d4-a6b2-11e5-84af-5861aae60b0c.png)

**Comment by MasterJames** [comment]
I am impressed. Makes one wanna reach out and touch it, or turn the knob as it were.

**Comment by borismus** [comment]
Awesome. Got a link to the demo?

**Comment by FreakTheMighty** [comment]
@mrdoob I'v modified webvr-boilerplate with the changes you suggested in https://github.com/mrdoob/three.js/issues/7732#issuecomment-166032503 and grabbed the latest from dev, but I'm still only seeing one eye rendered. The effect is the same as you saw here https://github.com/mrdoob/three.js/issues/7732#issuecomment-163850979.

The Cardboard demo you have in dev looks good on desktop, but I can't get it to run on my Nexus 5. I get a very playful, "Rats! WebGL hit a snag."

Can anyone else on this thread confirm the fix on mobile? Also, @brianchirls did you ever get webvr polyfill working with this fix?

**Comment by mrdoob** [comment]
> The Cardboard demo you have in dev looks good on desktop, but I can't get it to run on my Nexus 5. I get a very playful, "Rats! WebGL hit a snag."

Yes, that's because the shadow code is a hack at the moment (fixing it right now). Try removing the shadow from the example.

**Comment by brianchirls** [comment]
@FreakTheMighty No, I'm still having trouble with it. Will try to debug over the weekend and report an update.

**Comment by brianchirls** [comment]
@mrdoob I can confirm what @FreakTheMighty [said a couple days ago](https://github.com/mrdoob/three.js/issues/7732#issuecomment-171839079), that only one eye is being rendered. And now with or without shadows.

Here's what's happening (for each eye):
1. webvr-boilerplate sets the viewport (and scissor) to half the screen
2. webvr-boilerplate calls `WebGLRenderer.render` with a `renderTarget`
3. `WebGLRenderer.render` calls `setRenderTarget`, which then sets the viewport to the full size of the `renderTarget`, overwriting the previous call to set the viewport.

Also, `WebGLRenderer.setViewport` does not take into account the pixel ratio. It looks like that's only considered inside of `setRenderTarget` when the `renderTarget` is `null`.

**Comment by mrdoob** [comment]
Yes, `webvr-boilerplate` needs to be updated.

They should use these files as reference:

http://rawgit.com/mrdoob/three.js/dev/examples/webgl_effects_cardboard.html
http://rawgit.com/mrdoob/three.js/dev/examples/webgl_effects_cardboard.html?0 (no shadows)
http://rawgit.com/mrdoob/three.js/dev/examples/js/effects/CardboardEffect.js

**Comment by borismus** [comment]
The point of webvr-boilerplate is that it's compatible with WebVR via
VREffect. This is not true for CardboardEffect.

So it will take a bit of thinking to integrate @mrdoob's faster
implementation into
https://github.com/borismus/webvr-boilerplate/blob/master/src/cardboard-distorter.js
.

**Comment by jzitelli** [comment]
@brianchirls @FreakTheMighty if you are having problems, you can try also my quick hack of CardboardEffect into webvr-boilerplate: http://jzitelli.github.io/webvr-boilerplate/cardboardeffect/index.html

Also be aware that the eye separation for the StereoCamera which is currently in use for CardboardEffect is computed based on this formula: stereoCamera.focalLength / 30 * 0.5, that was throwing me off for a bit.

**Comment by brianchirls** [comment]
@borismus @mrdoob: We should probably try to coordinate a bit here. If we can have at least a branch of boilerplate with the r74dev implemented, we'll be able to test to make sure it's working. I'd hate to see r74 get released only to find out later about some VR-breaking bug in it that won't be patched until the next release.

@mrdoob: do we have a release timeline for r74?

Thanks, everyone.

**Comment by brianchirls** [comment]
@borismus: If you'd like some help with that integration, I'm happy to do what I can. Find me over at the webvr slack.

**Comment by mrdoob** [comment]
> Also be aware that the eye separation for the StereoCamera which is currently in use for CardboardEffect is computed based on this formula: stereoCamera.focalLength / 30 * 0.5, that was throwing me off for a bit.

Would you suggest a better default?

**Comment by jzitelli** [comment]
@mrdoob it seems to me that the projection matrices set up in StereoCamera are suitable for stereo images that would be viewed through 3D glasses with non-distorting lenses, glasses which don't fill your whole field of view and the goal is not necessarily to make the viewer feel immersed in the 3D environment (the 3D image will not be filling their whole field of vision if non-distorting lenses are being used), e.g. creating images for a 3D movie that would be watched in a movie theater.

For VR, where we view the screen through distorting lenses, the analysis in http://paulbourke.net/stereographics/stereorender/ would not apply, since the paths to object / screen / eyes are no longer simple lines... And we want the virtual camera separation to match the physical separation of our eyes / the viewer lenses... that is my understanding, I'm not certain if I am correct though :)

My suggestion is to continue using VREffect (which implements the projection matrices based on the given eye separation and other factors), but augment it with the option of rendering the left/right images to distortion meshes as in CardboardEffect.  This is what I'm attempting, I'm happy to collaborate/coordinate as well!  I'm subvr on webvr slack.

**Comment by mrdoob** [comment]
`VREffect` uses the projection matrices that the browser gets from the HMD in use, so `VREffect` works only when `WebVR` is available. For `CardboardEffect` we need to compute our own.

Anyway, have you tried to just increase the `fov` of the camera?

**Comment by brianchirls** [comment]
My concern here is that `setViewport` and `setScissor` do not work as documented, since whatever values you call them with will be ignored as soon as you try to render to another render target...*and* because they ignore the pixel ratio.

If you'd like to change the way those methods work, then so be it, but it should be reflected in the documentation and they should be replaced with a way to accomplish the same thing.

If there were a clear, functioning way set those properties, then I'm sure @borismus et al could find a way to make the polyfill work. And I'm concerned this issue may cause other problems we're not seeing yet.

**Comment by jzitelli** [comment]
@mrdoob sorry, I think I'm going off track / misleading by talking about projection matrices... it's the eye / camera separation that I'm concerned about - I don't think this should be a degree of freedom that is indirectly controlled via the focalLength property of PerspectiveCamera.  I think it should be set according to the Cardboard lens separation and the scale of the three.js scene.  Cardboard V1 lens separation: 0.06 meters, V2: 0.064 meters (according to info in webvr-boilerplate device-info.js)

My suggestion about augmenting / continuing to use VREffect was more focused on the effort to port the distortion mesh back into webvr-boilerplate.  I can definitely see the value in also having CardboardEffect which does not rely on WebVR API or the WebVR polyfill to be used.

**Comment by mrdoob** [comment]
> My concern here is that `setViewport` and `setScissor` do not work as documented, since whatever values you call them with will be ignored as soon as you try to render to another render target...*and* because they ignore the pixel ratio.

I fixed this last month already (check my Dec 19 comment).

**Comment by mrdoob** [comment]
@jzitelli Ah, I think I understand now. So you're suggesting to change `var eyeSep = focalLength / 30 * 0.5;` to `var eyeSep = 0.064;`?

**Comment by brianchirls** [comment]
@mrdoob: I'm looking at your Dec 19 comments, and I don't see a solution. Perhaps you could clarify?

If you're talking about calling `setSize`, I can't get it to work and don't see how it could. We still need the full width of two eyes in the distortion renderTarget, so changing the size doesn't help, *and* the horizontal offset is still overwritten anyway.

Again, `setViewport` and `setScissor` are now broken and if they are not going to be fixed or if they are only for internal use, then that should be reflected in the documentation, along with an explanation of an alternative approach.

**Comment by brianchirls** [comment]
Please hold. I was confused here, because the links to the code you posted are wrong now that other changes have made to the code.

It's here now:
https://github.com/mrdoob/three.js/blob/67522a7663c159f2a05aba8c12e109d3bad61a76/examples/js/effects/CardboardEffect.js#L114-L125

Let me see what I can work out and then I'll get back to you.

**Comment by mrdoob** [comment]
@brianchirls http://rawgit.com/mrdoob/three.js/dev/examples/webgl_effects_cardboard.html shows that the issue you reported (stereo breaking when using shows) is now fixed.

**Comment by brianchirls** [comment]
@mrdoob Yes, I understand. I just want to be able to confirm for certain that it's working in a general case. There are some other smaller issues that this is turning up. If you prefer, I can file them separately.

I'm going through your code and am almost able to replicate your solution with my own fork of webvr-boilerplate. There is one snag that maybe you could help me with? In my code the right eye is doing a clear of the full render target, obliterating everything that was just rendered for the left eye. What are you doing to prevent this from happening in your example?

Thanks.

**Comment by jzitelli** [comment]
@mrdoob yes, that's really all I was trying to say, heh

**Comment by mrdoob** [comment]
@jzitelli apart from the eye separation, all the rest of the code (projection matrix computation, ...) looks good to you?

**Comment by jzitelli** [comment]
oh sorry, i think that value should be halved (0.064 is the total distance from lens center to lens center, i.e. translate one camera 0.032 left, the other 0.032 right - using the variable name eyeSep there is misleading, maybe call it halfEyeSep?)... the rest looks ok to me, but I'm not an expert, forget what I said before re: projections... I am taking some time to research and understand how the cameras should be set up for Cardboard, maybe others here can give better feedback.

I think it would be good to set up a benchmark scene (e.g. something like the Oculus configuration scene - a desk with various common objects, it provides a good subjective test of how immersive / correctly scaled the VR is).

**Comment by mrdoob** [comment]
@jzitelli fixed! 😅
